aten/src/ATen/native/mkldnn/xpu/detail/Utils.h

```diff
@@ -0,0 +1,376 @@
+  auto block_size = 1;
+  // const auto& blk = md->format_desc.blocking;
+  dnnl_dims_t md_inner_blks;
+  dnnl_dims_t md_blk_inner_idxs;
+  dnnl_memory_desc_query(md, dnnl_query_inner_idxs, &md_blk_inner_idxs);
+  dnnl_memory_desc_query(md, dnnl_query_inner_blks, &md_inner_blks);
+  for (int iblk = 0; iblk < md_inner_nblks; ++iblk) {
+    blocks[md_blk_inner_idxs[iblk]] *= md_inner_blks[iblk];
+    block_size *= md_inner_blks[iblk];
+  }
+
+  // A custom comparator to yield linear order on perm
+  auto idx_sorter = [&](const int a, const int b) -> bool {
+    if (strides[a] == strides[b] && md_padded_dims[a] == md_padded_dims[b])
+      return a < b;
+    else if (strides[a] == strides[b])
+      return md_padded_dims[a] < md_padded_dims[b];
+    else
+      return strides[a] < strides[b];
+  };
+  std::sort(perm, perm + md_ndims, idx_sorter);
+
+  auto min_stride = block_size;
+  for (int idx = 0; idx < md_ndims; ++idx) {
+    const int d = perm[idx];
+
+    // Make an exception for strides[d] == 0 as it has broadcast semantics
+    // Note: owing to being sorted, these are the initial strides
+    if (strides[d] == 0)
+      continue;
+    else if (strides[d] < min_stride)
+      return false;
+
+    // update min_stride for next iteration
+    const auto padded_dim = *md_padded_dims[d];
+    min_stride = block_size * strides[d] * (padded_dim / blocks[d]);
+  }
+  return true;
+}
+
+static inline bool is_broadcast(const at::Tensor& t) {
+  for (int i = 0; i < t.dim(); i++) {
+    if (t.stride(i) == 0)
+      return true;
+  }
+  return false;
+}
+
+static inline bool is_onednn_matmul_strides(
+    const at::Tensor& tensor,
+    bool is_dst = false) {
+  // https://oneapi-src.github.io/oneDNN/dev_guide_matmul.html
+  // oneDNN matmul only support 2-dim and 3-dim
+  // 2D src(Mxk), wei(KxN), dst(MxN)
+  // 3D src(SxMxK), wei(WxKxN), dst(DxMxN)
+  auto sizes = tensor.sizes();
+  auto tensor_dim = sizes.size();
+  if (tensor_dim != 2 && tensor_dim != 3)
+    return false;
+
+  if (tensor.is_contiguous())
+    return true;
+
+  // the overlaped cases are not supported
+  dnnl::memory::dims strides = get_onednn_strides(tensor);
+  int64_t storage_size = 1;
+  for (size_t dim = 0; dim < tensor_dim; ++dim)
+    storage_size += (sizes[dim] - 1) * strides[dim];
+  if (storage_size < tensor.numel())
+    return false;
+
+  // the broadcast cases are not supported
+  if (is_broadcast(tensor)) {
+    return false;
+  }
+
+  if (is_dst) {
+    // The memory format of the destination tensor should always
+    // be plain with n axis contiguous
+    if (strides[-1] != 1)
+      return false;
+  } else {
+    // the src and weight must have at least one of the axes
+    // m or k and n or k contiguous (i.e., stride=1) respectively.
+    if (strides[tensor_dim - 1] != 1 && strides[tensor_dim - 2] != 1)
+      return false;
+  }
+
+  if (!onednn_strides_check(tensor))
+    return false;
+
+  return true;
+}
+
+static bool is_wrapped_number(const at::Tensor& t) {
```

Comment:
remove this util, call the pytorch API directly, should be more clear for maintainer?

Reply:
removed